### ISSUE TITLE
Bug: `euler_characteristic` doesn't give right result for meshes with boundary

### ISSUE THREAD

**Issue by keenancrane** [libigl/libigl#2249]
During an SGI project, we discovered that the `euler_characteristic` method does not return the right result for meshes with boundary.  The standard Euler characteristic formula is

V - E + F + B = (2-2g) = χ,

but the current implementation of `euler_characteristic` does not compute the number of boundary loops.  This number could be computed via the `boundary_loops` method.

Compare with the Geometry Central implementation found here:

https://github.com/nmwsharp/geometry-central/blob/071756157a6122ead8fa978a35b06d1acc16451f/src/surface/manifold_surface_mesh.cpp#L515

**Comment by keenancrane** [comment]
Yeah, you’re right—the definition you cite also agrees with the definition based on Betti numbers. (Perhaps I should file an issue on Geometry Central instead…)

The practical issue is that the Euler characteristic alone is ambiguous when trying to perform some reasonably common tasks. For instance a double torus and a disk with two holes both have characteristic -1. In the context of the SGI project, students were trying to extract patches that can be flattened, and checking the Euler characteristic alone was giving them the wrong answer (patches could have handles).

It looks like you can get the number of handles h by calling `euler_characteristic` to get e, then `boundary_loops` to get b; h is then 1-(e+b)/2. But this is pretty indirect, requiring a bit of “expert knowledge.”

So, perhaps a more constructive suggestion for this is to make a feature request:

> **Request:** add a method `handles` that returns the number of topological handles, for meshes with or without boundary.

(Will change the issue title also.)

Thanks!

**Comment by alecjacobson** [comment]
That's a nice feature request. I have that in matlab gptoolbox's [statistics](https://github.com/alecjacobson/gptoolbox/blob/46a26ea5ab5dfccf4da39385381ca986f66bacfd/mesh/statistics.m#L188) and in C++ [in gp-cli](https://github.com/alecjacobson/gp-cli/blob/28f042d8f8e1df42881ce338b4bd28af051b0457/meshstatistics.cpp#L222). Might as well add it to libigl sometime.

**Comment by alecjacobson** [comment]
See also https://github.com/libigl/libigl/issues/1331#issuecomment-546644630
